packages/core/src/browser/connection-status-service.ts

```diff
@@ -144,9 +122,20 @@ export class FrontendConnectionStatusService extends AbstractConnectionStatusSer
     @inject(WebSocketConnectionProvider) protected readonly wsConnectionProvider: WebSocketConnectionProvider;
     @inject(PingService) protected readonly pingService: PingService;
 
+    constructor(@inject(ConnectionStatusOptions) @optional() protected readonly options: ConnectionStatusOptions = ConnectionStatusOptions.DEFAULT) {
+        super(options);
+    }
+
     @postConstruct()
     protected init(): void {
-        this.schedulePing();
+        this.wsConnectionProvider.onSocketOpened(() => {
+            this.updateStatus(true);
+            this.schedulePing();
+        });
+        this.wsConnectionProvider.onSocketClosed(() => {
+            this.clearTimeout(this.scheduledPing);
+            this.updateStatus(false);
+        });
         this.wsConnectionProvider.onIncomingMessageActivity(() => {
```

Comment:
Given this, with this change we update the status when a connection is opened, when the connection closes, and when we receive messages. I could not find a place where `schedulePing` would timeout though. It will set the status to `false` when an error occurs, but that doesn't account for timeouts IIUC?

Reply:
>It will set the status to false when an error occurs, but that doesn't account for timeouts IIUC?

Yeah, that's the point of this PR. Because timeout doesn't indicate that we have a broken connection, which is displayed as "offline" mode. The main problem is that when "offline" mode is triggered by timeout, the ping request receives a response from the server, then "online" status is triggered immediately after ping request. And this is quite false positive displaying the status, because user is able to call any other requests, e.g. create file, open it in the editor, open the terminal and so on.